File sync for devfiles

johnmcollier opened the pull request:
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:

/kind feature


**What does does this PR do / why we need it**:
This PR implements file syncing for devfile components, using the kubernetes client library. The push code is based on the existing push/sync code used in classic odo, with only a few small tweaks required.

Overall this PR does the following:
- It updates the Devfile deployment code to create a shared emptyDir volume (called `odo-projects`) and mount it under `/projects` in all of the devfile components that set `mountSources: true`
    - This is in line with what Che does for mounting project source
- Creates a `Push()` function in the devfile adapters to handle file syncing
    - If syncing to the pod for the first time (or if `--force` is set), it will sync all of files in the source directory using the `sync` package
    - If syncing to an already running component:
       - it syncs only the files that changed (by checking odo-file-index.json)
       - If any files were deleted from the source directory, it deletes from the remote pod.
- Creates a `DoesComponentExist()` function in the devfile adapters
    - Returns true if a component with the specified label already exists
    - Wrapper for `utils.ComponentExists`
- Moved `IsEmpty` into the `pkg/utils` package from `pkg/component`
    - Needed for both odo "classic" and "devfile" components, so needed to move it to a neutral location.

- Tests
    - Unit tests for new functions
         - Can't really unit test Push(), as it can't be mocked with the fake Kube client (e.g. requires exec'ing to a running pod)
    - Integration tests to cover push and sync scenarios

**Which issue(s) this PR fixes**:

Fixes #2473, https://github.com/openshift/odo/issues/2673

**How to test changes / Special notes to the reviewer**:
1. Build my branch and run `make`
2. Run `make test` to verify unit tests pass
3. Enable experimental mode
4. Run `odo push` verify that the deployment starts and files get synced
5. Change/add/remove files, verify the proper changes get reflected on the container.

amitkrout commented:
/retest

johnmcollier commented:
/retest

johnmcollier commented:
Updated to include:
`-namespace` flag on odo push
    - Unrelated to file syncing, but was a requirement to implement integration tests
- integration tests 🎉 
    - tests to verify both odo push and syncing works
    - Adds two simple devfiles, both nodejs. One single container, one multi container
    - Just a few simple tests right now as there isn’t a complete end to end scenario.

johnmcollier commented:
/retest

amitkrout commented on .travis.yml:
```diff
@@ -98,6 +98,7 @@ jobs:
         - travis_wait make test-cmd-app
         - travis_wait make test-cmd-push
         - travis_wait make test-cmd-project
+        - travis_wait make test-cmd-devfile-push
```
+1,
FYI - Travis job runs test on 3.11 cluster

amitkrout replied:
Update the [name:](https://github.com/openshift/odo/pull/2681/files#diff-354f30a63fb0907d4ad57269548329e3L90) field accordingly. For example 
```
name: "watch, storage, app, project, push and devfile push command integration tests"
```

johnmcollier replied:
Good catch, updated!

amitkrout commented on Makefile:
```diff
@@ -190,6 +190,11 @@ test-cmd-pref-config:
 test-cmd-push:
 	ginkgo $(GINKGO_FLAGS) -focus="odo push command tests" tests/integration/
 
+# Run odo push command tests
```
```Run odo push command tests``` -> ```Run odo push devfile command tests```

johnmcollier replied:
Thanks, updated!

amitkrout commented on tests/integration/cmd_devfile_push_test.go:
```diff
@@ -0,0 +1,156 @@
+		context = helper.CreateNewContext()
+		currentWorkingDirectory = helper.Getwd()
+		helper.Chdir(context)
+		os.Setenv("ODO_EXPERIMENTAL", "true")
```
Do not use it like this because odo test spec (It) runs on two ginkgo test node in parallel, so it may cause race condition. To avoid such situation we have a env variable which can control preference path.

You can use 
```
os.Setenv("GLOBALODOCONFIG", filepath.Join(context, "config.yaml"))
```
and in the spec i mean in the It block you can update the preference value. For example
```
helper.CmdShouldPass("odo", "preference", "set", "Experimental", "true")
```

johnmcollier replied:
I was wondering about the tests running in parallel and whether or not that would cause problems, good to know! I'll update my PR to include that.

amitkrout commented on tests/integration/cmd_devfile_push_test.go:
```diff
@@ -0,0 +1,156 @@
+		helper.DeleteProject(project)
+		helper.DeleteDir(context)
+		helper.Chdir(currentWorkingDirectory)
+		os.Unsetenv("ODO_EXPERIMENTAL")
```
As per https://github.com/openshift/odo/pull/2681/files#r388746348 use ```os.Unsetenv("GLOBALODOCONFIG")``` instead.

amitkrout commented on tests/integration/cmd_devfile_push_test.go:
```diff
@@ -0,0 +1,156 @@
+			output := helper.CmdShouldPass("odo", "push", "--devfile", "devfile.yaml", "--namespace", project)
+			Expect(output).To(ContainSubstring("✓  Waiting for component to start"))
+			Expect(output).To(ContainSubstring("✓  Push devfile component"))
+			Expect(output).To(ContainSubstring("✓  Changes successfully pushed to component"))
```
One assertion that confirms the devfile push is successful would be enough. So  ```Expect(output).To(ContainSubstring("Changes successfully pushed to component"))``` would be the proper fit imo. Do not to pass character like ```✓``` in assert as it may behave differently on terminal like PowerShell.

johnmcollier replied:
Ah, right good point.

amitkrout commented on tests/integration/cmd_devfile_push_test.go:
```diff
@@ -0,0 +1,156 @@
+	// This is run after every Spec (It)
+	var _ = AfterEach(func() {
+		helper.DeleteProject(project)
+		helper.DeleteDir(context)
```
It will throw error on windows platform. Rearranging the line with the below one will fix the problem

```
helper.Chdir(currentWorkingDirectory)
helper.DeleteDir(context)
```

johnmcollier replied:
Ahh, that's a good point about Windows. Updated.

amitkrout commented on tests/integration/cmd_devfile_push_test.go:
```diff
@@ -0,0 +1,156 @@
+			Expect(output).To(ContainSubstring("No file changes detected, skipping build"))
+			helper.ReplaceString(filepath.Join(context, "server.js"), "node listening on", "UPDATED!")
+
+			helper.CmdShouldPass("odo", "push", "--context", context+"/nodejs-ex", "--namespace", project)
```
Use ```filepath.Join(context, "nodejs-ex")``` instead of ```context+"/nodejs-ex"```.

johnmcollier replied:
Updated, good catch!

johnmcollier commented:
@amitkrout Thanks for reviewing! I've addressed your review comments in a new commit.

johnmcollier commented:
/test v4.2-integration-e2e-benchmark

rajivnathan commented on pkg/devfile/adapters/common/interface.go:
```diff
@@ -1,6 +1,10 @@
 // ComponentAdapter defines the functions that platform-specific adapters must implement
 type ComponentAdapter interface {
 	Start() error
+	Push(path string, out io.Writer, files []string, delFiles []string, isForcePush bool, globExps []string, show bool) error
```
Actually, do we need both Start() and Push()? Why can't one function handle both? ie. start if the component doesn't already exist and then sync.

johnmcollier replied:
We could merge the two, but my thinking for keeping them separate was to keep things separate and to avoid having one big monolithic start function

johnmcollier commented on pkg/devfile/adapters/kubernetes/component/adapter.go:
```diff
@@ -62,11 +68,91 @@ func (a Adapter) Start() (err error) {
+
+// getFirstContainerWithSourceVolume returns the first container that set mountSources: true
+// If no container was found, that means there's no container to sync to, so return an error
+func getFirstContainerWithSourceVolume(containers []corev1.Container) (string, error) {
```
Yup, sounds good to me

johnmcollier commented on pkg/devfile/adapters/kubernetes/component/adapter.go:
```diff
@@ -62,11 +68,91 @@ func (a Adapter) Start() (err error) {
 
-	_, err = a.Client.WaitAndGetPod(watchOptions, corev1.PodRunning, "Waiting for component to start")
-	return err
+	return "", fmt.Errorf("No containers specified mountSources: true")
```
Makes sense to me.

johnmcollier commented on pkg/devfile/adapters/helper.go:
```diff
@@ -17,14 +17,32 @@ func NewPlatformAdapter(componentName string, devObj devfile.DevfileObj) (Platfo
 
-func createKubernetesAdapter(adapterContext common.AdapterContext) (PlatformAdapter, error) {
+// NewPlatformAdapter returns a Devfile adapter for the targeted platform
+func NewPlatformAdapterWithNamespace(componentName string, devObj devfile.DevfileObj, namespace string) (PlatformAdapter, error) {
```
Ooh, that's a better approach, I like it. I'll go take a look.

johnmcollier commented:
@rajivnathan Thanks for the review! I've addressed your review comments. I've also moved the `FilterIgnores` function into the `utils` package (and written unit tests for it!), as the refactoring of `Push()` into `Start()` required it.

johnmcollier commented:
/retest

johnmcollier commented:
/retest 🙃

johnmcollier commented:
/retest

johnmcollier commented:
/retest https://github.com/openshift/odo/issues/2642 is making me sad 😢

johnmcollier commented:
/retest

rajivnathan commented on pkg/devfile/adapters/kubernetes/component/adapter.go:
```diff
@@ -28,7 +36,100 @@ type Adapter struct {
 // Start updates the component if a matching component exists or creates one if it doesn't exist
-func (a Adapter) Start() (err error) {
+// Once the component has started, it will sync the source code to it.
+func (a Adapter) Start(path string, out io.Writer, ignoredFiles []string, forceBuild bool, globExps []string, show bool) (err error) {
```
`out` and `show` don't appear to be used, can we remove them?

johnmcollier replied:
They are used, in the call to `pushLocal`

johnmcollier replied:
Done

rajivnathan commented on pkg/devfile/adapters/interface.go:
```diff
@@ -1,5 +1,7 @@
+
 type PlatformAdapter interface {
-	Start() error
+	Start(path string, out io.Writer, ignoredFiles []string, forceBuild bool, globExps []string, show bool) error
```
I think we discussed renaming `Start` to `Push`. Should we still go ahead with that?

johnmcollier replied:
Done

kadel commented:
@johnmcollier 
All existing devfiles are expecting that the source code will be in `/projects/<project-name>`.
I've noticed that you are pushing files to `/projects/`. 

This will break every existing devfile there is :-(
For example, neither of this will work:
 - https://github.com/eclipse/che-devfile-registry/blob/master/devfiles/java-web-spring/devfile.yaml
 - https://github.com/elsony/devfile-registry/blob/master/devfiles/springBoot/devfile.yaml

I know that there is no good solution for us, as we can't work with multiple projects in devfile right now.

What if we do this:
- If there is exactly one project defined in devfile, odo pushes code to `/projects/<project-name>/` 
- In all other cases (no project, or multiple projects) it pushes the source code directly to `/projects`.

This will make most of the devfiles usable with odo.

rajivnathan commented:
@kadel I agree we need to figure out where source should be synced in a way that works for most devfiles. I've just finished trying out John's latest changes and things work pretty well so maybe we can open a separate issue to handle deciding where files should be synced? We can then go through the use cases in a bit more detail and updating the code to change where to sync should be pretty simple.

johnmcollier commented:
@kadel @rajivnathan The changes to support pushing to `/projects/<projectName>` when exactly one project is defined was pretty straightforward, so I've pushed up a commit that does the following:
- Before syncing the source code, odo checks how many projects were defined in the devfile
     - If exactly one, it sets the "sync folder" to `/projects/<projectName>`, ensures `/projects/<projectName>` exists in the container and pushes the files to that folder
     - If either no projects are defined, or multiple projects are defined, odo syncs to `/projects`

As Rajiv notes, we'll probably need to do some more more work later to properly iron out support for multiple projects, but this is a fine start at the very least.

rajivnathan commented on pkg/devfile/adapters/kubernetes/component/adapter.go:
```diff
@@ -246,6 +248,22 @@ func (a Adapter) pushLocal(path string, files []string, delFiles []string, isFor
 
+	// If there's only one project defined in the devfile, sync to `/projects/project-name`, otherwise sync to /projects
+	projects := a.Devfile.Data.GetProjects()
+	fmt.Println(len(projects))
```
Looks like an errant println slipped through

johnmcollier replied:
That was an errant debugging line that slipped through, removed.

rajivnathan commented:
/lgtm opened https://github.com/openshift/odo/issues/2705 for follow up discussion on how best to handle multiple projects

girishramnani commented on pkg/devfile/adapters/kubernetes/component/adapter.go:
```diff
@@ -117,11 +211,105 @@ func (a Adapter) Create() (err error) {
+}
+
+// Push syncs source code from the user's disk to the component
+func (a Adapter) pushLocal(path string, files []string, delFiles []string, isForcePush bool, globExps []string) error {
```
this function is doing alot of things, lets try to extract certain portions to make unit testing better. 
two suggestions would be be -
1 - https://github.com/openshift/odo/pull/2681/files#diff-825a1a88d75e789899c58e89a79eb767R254-R264 can be extracted into a function
2 - https://github.com/openshift/odo/pull/2681/files#diff-825a1a88d75e789899c58e89a79eb767R268-R277 can be extracted into a function

Also if the function has more then 5 args please try to create a struct to pass on data or consider moving it as part of the method's struct if it doesn't change through out the lifecycle of the struct. 
e.g. path and ignoredFiles fields might not change throughout the life of `Adapter` so maybe it can be places as part of the Adapter's struct? - this is just an example.

johnmcollier replied:
@girishramnani I'm not sure moving those sections into separate functions would make unit testing easier - they both involve exec'ing into an already running container and manipulating the files that exist there, and that can't be mocked with the fake kube client.

girishramnani replied:
we can mock it without the fake kube client and it will definitely make unit testing pushLocal easier because when you would write test for pushLocal you would mock the https://github.com/openshift/odo/pull/2681/files#diff-825a1a88d75e789899c58e89a79eb767R254-R264 and https://github.com/openshift/odo/pull/2681/files#diff-825a1a88d75e789899c58e89a79eb767R268-R277 functions.

johnmcollier replied:
I think what makes it different is that we're interacting with the kube client different than  we do for other resources, we're interacting directly with the rest client, not calling an existing `exec` function:
```
	req := c.KubeClient.CoreV1().RESTClient().
		Post().
		Namespace(c.Namespace).
		Resource("pods").
		Name(podName).
		SubResource("exec").
		VersionedParams(&podExecOptions, scheme.ParameterCodec)
```
and
```
	exec, err := remotecommand.NewSPDYExecutor(config, "POST", req.URL())
	if err != nil {
		return errors.Wrapf(err, "unable execute command via SPDY")
	}
	// initialize the transport of the standard shell streams
	err = exec.Stream(remotecommand.StreamOptions{
		Stdin:  stdin,
		Stdout: stdout,
		Stderr: stderr,
		Tty:    tty,
	})
```

(see https://github.com/openshift/odo/blob/master/pkg/kclient/pods.go#L93)

johnmcollier replied:
As a compromise, rather than putting those sections with the exec into separate functions, what about moving everything _before_ the exec into their own functions (maybe `getCommandToCreateFolder` and `getCommandToDelete`)? Then we could still write unit

(Note when moving `ExecCMDInContainer` over from occlient, I couldn't find any references in the existing odo code that had unit tests with it)

girishramnani replied:
> (Note when moving `ExecCMDInContainer` over from occlient, I couldn't find any references in the existing odo code that had unit tests for it, or unit tests for functions that called it, so other parts of odo have had this problem too.)

and that is the reason that we should not do this mistake when we have the opportunity to re-write/replace this functionality

girishramnani replied:
> As a compromise, rather than putting those sections with the exec into separate functions, what about moving everything _before_ the exec into their own functions (maybe `getCommandToCreateFolder` and `getCommandToDelete`)? Then we could still write unit tests.

I keep that upto you, I just wanted to break this function down into smaller pieces to make it more unit testable

johnmcollier replied:
@girishramnani I'll go with approach for now (and see if there are any other pieces we can break up). I'll also address your comment regarding the number of parameters.

Since unit tests for the `ExecCmdInContainer` function will require more work / intevestigation / design, I'll address it as part of https://github.com/openshift/odo/issues/2652 (which is also dealing with unit-testing exec)

kadel commented:
```
▶ odo push
 •  Push devfile component spring-petclinic  ...
 ✓  Checking file changes for pushing [12ms]
 ✓  Waiting for component to start [121ms]
 ✓  Syncing files to the component [2s]
 ✓  Push devfile component spring-petclinic [3s]
 ✓  Changes successfully pushed to component

▶ odo push
 •  Push devfile component spring-petclinic  ...
 ✓  Checking file changes for pushing [8ms]
 ✓  Waiting for component to start [114ms]
 ✗  Syncing files to the component [30s]
 ✗  Failed to start component with name spring-petclinic.
Error: Failed to create the component: Failed to sync to component with name spring-petclinic: error while streaming command: Internal error occurred: error executing command in container: container is not created or running
```

For the second push, I changed the environment variable in devfile.
I guess it's because of this https://github.com/openshift/odo/issues/2661

johnmcollier commented:
@kadel Yup, looks like it

kadel commented:
> @kadel Yup, looks like it, since changing the environment variable would have restarted the deployment.

Push command needs to detect this and wait, till the pod is ready again. 
This can be addressed in a separate PR, but we need to fix it.

kadel commented:
/approve

johnmcollier commented:
Separated out pushLocal into 3 more functions and added unit tests for them. Going to refactor ExecCMDInContainer in a separate PR to make it:
- Take fewer params (probably a struct of some kind in the sync package). Will also update pushlocal and push to take fewer params too.
- Make it more unit-testable.

I'll refactor both ExecCMDInContainer in kclient and occlient (as the syncclient interface requires clients to implement this function)

elsony commented:
Confirmed with @johnmcollier this sync support do incremental sync to only sync the changed file.
